File: backend/internal/domain/pagination.go
package domain

import "math"

type PaginatedResponse[T any] struct {
	Data        []T   `json:"data"`
	Page        int   `json:"page"`
	PageSize    int64 `json:"page_size"`
	TotalItems  int64 `json:"total_items"`
	TotalPages  int   `json:"total_pages"`
	HasNextPage bool  `json:"has_next_page"`
	HasPrevPage bool  `json:"has_prev_page"`
}

type PaginationParams struct {
	Page     int
	PageSize int64
}

func BuildPaginationParams(page int, pageSize, totalItems int64) PaginationParams {
	if pageSize == -1 {
		pageSize = totalItems
	}

	return PaginationParams{Page: page, PageSize: pageSize}
}

func (p PaginationParams) Offset() int64 {

	return int64(p.Page-1) * p.PageSize
}

func CreatePaginatedResponse[T any](data []T, totalItems int64, paginationParams PaginationParams) *PaginatedResponse[T] {
	totalPages := int(math.Ceil(float64(totalItems) / float64(paginationParams.PageSize)))

	return &PaginatedResponse[T]{
		Data:        data,
		Page:        paginationParams.Page,
		PageSize:    paginationParams.PageSize,
		TotalItems:  totalItems,
		TotalPages:  totalPages,
		HasNextPage: paginationParams.Page < totalPages,
		HasPrevPage: paginationParams.Page > 1,
	}
}

type RealEstatePaginatedResponse = PaginatedResponse[*RealEstatePropertyInput] // para swagger
type NoBodyResponse struct{}                                                   // para swagger
type ErrorResponse struct {                                                    // para swagger
	Error string `json:"error"`
}

File: frontend/lib/data/models/pagination.dart

import '../../domain/entity/pagination.dart';

class PaginationModel<T> extends PaginationEntity<T> {

  PaginationModel({
    required super.data,
    required super.page,
    required super.pageSize,
    required super.totalItems,
    required super.totalPages,
    required super.hasNextPage,
    required super.hasPrevPage,

});

  factory PaginationModel.fromJson(Map<String, dynamic> json,
      T Function(Map<String, dynamic>) creator,) {

    return PaginationModel(
      data: (json['data'] as List)
          .map((item) => creator(item as Map<String, dynamic>))
          .toList(),
      page: json['page'] ?? 0,
      pageSize: json['page_size'] ?? 0,
      totalItems: json['total_items'] ?? 0,
      totalPages: json['total_pages'] ?? 0,
      hasNextPage: json['has_next_page'] ?? false,
      hasPrevPage: json['has_prev_page'] ?? false,
    );
  }




}